### DISCUSSION TITLE
Multi line text editor decoration using vscode.window.createTextEditorDecorationType()

### DISCUSSION THREAD

**Discussion by Pranith12250** [microsoft/vscode-discussions#1202] [Extension Development QnA] [answered]
I am trying to create a simple extension. This extension displays some text on an active text editor as ghost text only when the editor is empty, i.e., there is no text typed into the editor. But the moment something is typed in the editor, the ghost text must disappear and once again reappear when anything that has been typed is deleted out.

I do not have an extensive knowledge on ways to achieve this, so I used text decoration to achieve this. For this, I wrote the below code inside the activate() function:

```
const placeholderText='Hi! The extension is working!\n\n This should be the third line\n And this the fourth\nAnd this the fifth\nBye bye!\n\nThis text will automatically disappear as you start typing!';

const placeholderDecorationType = vscode.window.createTextEditorDecorationType({
        after: { contentText: placeholderText, color: new vscode.ThemeColor('editorCodeLens.foreground') },
    });

const updatePlaceholderDecorations = () => {
        vscode.window.visibleTextEditors.forEach((editor) => {
            const decorations: vscode.DecorationOptions[] = [];
            if (editor.document.getText().trim().length === 0) {
                decorations.push({ range: new vscode.Range(0, 0, 0, 0) });
            }
            editor.setDecorations(placeholderDecorationType, decorations);
        });
    };

const disposable1 = vscode.window.onDidChangeVisibleTextEditors(updatePlaceholderDecorations);
const disposable2 = vscode.workspace.onDidChangeTextDocument((event) => {
    if (event.contentChanges.length > 0) {
        updatePlaceholderDecorations();
    }
});

context.subscriptions.push(disposable1, disposable2);
```

I am expecting the text to appear over multiple lines but the final output seems to appear only in the first line as shown below:

<img width="1436" alt="Screenshot 2024-05-27 at 8 10 15 PM" src="https://github.com/microsoft/vscode-discussions/assets/127399614/7bbb8a33-279e-4003-9f8d-f5248e91e01e">

And this text disappears on typing anything as expected and returns when the typed text is removed as shown below:

<img width="1440" alt="Screenshot 2024-05-27 at 8 16 56 PM" src="https://github.com/microsoft/vscode-discussions/assets/127399614/06a68bb0-bcc5-4392-9c11-12214c3a662b">

The font style and colour and everything is just as I want it to be but I just want the text to appear over multiple lines. I have demonstrated the expected result below by manually typing it but I want this to appear in the way it looks in the previous screenshots attached above:

<img width="1440" alt="Screenshot 2024-05-27 at 8 18 57 PM" src="https://github.com/microsoft/vscode-discussions/assets/127399614/e56a427e-3ce4-4615-9e5d-a20ba4768f08">

So, please let me know the following:

1. What modification do I need to make in order to get the text decoration over multiple lines instead of a single line?
2. Or, is there a better way to achieve this same functionality by a different method. If so, please let me know how.
3. Also, I noticed that the decoration text does not appear until I actually kill the integrated terminal that automatically appears? Closing the terminal alone is not enough and I have to kill the terminal for the text to appear. Is there a way to solve this?

**Comment by isidorn** [comment, marked as answer]
These are super specific questions for your extension.
I suggest to try to break them down into smaller questions, and try to ask them in a more generlised fashion, so other extension authors might also benefit, and might share their experience.

If that does not help, then I suggest to try asking GH Copilot - should be able to pick up your context, and look at vscode.d.ts to help you achieve what you want.
